core/runtime/module_repository.hpp

```diff
@@ -10,18 +10,42 @@
 
 #include <gsl/span>
 
+#include "host_api/host_api.hpp"
+#include "module_instance.hpp"
 #include "outcome/outcome.hpp"
 #include "primitives/block_data.hpp"
 
-namespace kagome::runtime {
-  class RuntimeCodeProvider;
-}
-
 namespace kagome::runtime {
 
   class ModuleInstance;
   class Module;
   class Memory;
+  class RuntimeCodeProvider;
+
+  class BorrowedRuntimeInstance {
+   public:
+    BorrowedRuntimeInstance() = default;
+    BorrowedRuntimeInstance(std::shared_ptr<ModuleInstance> instance,
+                            std::function<void()> cache_release = {})
+        : instance_{std::move(instance)},
+          cache_release_{std::move(cache_release)} {}
+    ~BorrowedRuntimeInstance() {
+      if (cache_release_) {
+        cache_release_();
+      }
+    }
+    bool operator==(std::nullptr_t) {
+      return instance_ == nullptr;
+    }
+    ModuleInstance *operator->() {
+      return instance_.get();
+    }
+
+    std::shared_ptr<ModuleInstance> instance_;
+
+   private:
+    std::function<void()> cache_release_;
```

Comment:
You can typedef this, otherwise refactoring (which is quite likely) will be unpleasant.